File: packages/shared-types/src/canonical.ts
import { z } from 'zod';
import { ChannelSchema } from './identity.js';

// ── Canonical content variants ───────────────────────────────────────────

export const TextContentSchema = z.object({
  type: z.literal('text'),
  text: z.string().min(1).max(4096),
});

export const VoiceContentSchema = z.object({
  type: z.literal('voice'),
  media_url: z.string().url(),
  duration_seconds: z.number().int().nonnegative().optional(),
  transcription: z.string().nullable().optional(),
  transcription_language: z.string().optional(),
});

export const ImageContentSchema = z.object({
  type: z.literal('image'),
  media_url: z.string().url(),
  caption: z.string().optional(),
  analysis: z.string().nullable().optional(),
});

export const DocumentContentSchema = z.object({
  type: z.literal('document'),
  media_url: z.string().url(),
  filename: z.string().optional(),
  mime_type: z.string().optional(),
});

export const LocationContentSchema = z.object({
  type: z.literal('location'),
  lat: z.number(),
  lng: z.number(),
  name: z.string().optional(),
  address: z.string().optional(),
});

export const InteractiveReplyContentSchema = z.object({
  type: z.enum(['button_reply', 'list_reply']),
  selected_id: z.string(),
  selected_title: z.string(),
  context_message_id: z.string().optional(),
});

export const TemplateContentSchema = z.object({
  type: z.literal('template'),
  template_name: z.string(),
  language: z.string(),
  components: z.array(
    z.object({
      type: z.enum(['header', 'body', 'button']),
      parameters: z.array(
        z.object({
          type: z.enum(['text', 'image', 'document', 'currency', 'date_time']),
          value: z.string(),
        }),
      ),
    }),
  ),
});

export const OutboundButtonsContentSchema = z.object({
  type: z.literal('buttons'),
  body_text: z.string().min(1),
  header_text: z.string().optional(),
  footer_text: z.string().optional(),
  buttons: z.array(z.object({ id: z.string(), title: z.string().max(20) })).min(1).max(3),
});

export const OutboundListContentSchema = z.object({
  type: z.literal('list'),
  body_text: z.string().min(1),
  header_text: z.string().optional(),
  footer_text: z.string().optional(),
  button_text: z.string().max(20).default('Choose'),
  list_sections: z.array(
    z.object({
      title: z.string(),
      items: z
        .array(
          z.object({
            id: z.string(),
            title: z.string().max(24),
            description: z.string().max(72).optional(),
          }),
        )
        .min(1)
        .max(10),
    }),
  ),
});

export const CanonicalContentSchema = z.discriminatedUnion('type', [
  TextContentSchema,
  VoiceContentSchema,
  ImageContentSchema,
  DocumentContentSchema,
  LocationContentSchema,
  InteractiveReplyContentSchema,
  TemplateContentSchema,
  OutboundButtonsContentSchema,
  OutboundListContentSchema,
]);
export type CanonicalContent = z.infer<typeof CanonicalContentSchema>;

// ── Canonical message ────────────────────────────────────────────────────

export const DeliveryStatusSchema = z.enum(['queued', 'sent', 'delivered', 'read', 'failed']);
export type DeliveryStatus = z.infer<typeof DeliveryStatusSchema>;

export const CanonicalMessageSchema = z.object({
  message_id: z.string().min(1),
  direction: z.enum(['inbound', 'outbound']),
  channel: ChannelSchema,
  tenant_id: z.string().uuid().nullable(),
  thread_id: z.string().uuid().nullable(),
  sender_principal_id: z.string().uuid().nullable(),
  sender_identifier: z.string(),
  recipient_identifier: z.string(),
  timestamp: z.string().datetime(),
  content: CanonicalContentSchema,
  delivery: z
    .object({
      status: DeliveryStatusSchema,
      error: z.string().optional(),
      sent_at: z.string().datetime().optional(),
      delivered_at: z.string().datetime().optional(),
      read_at: z.string().datetime().optional(),
    })
    .optional(),
  raw_payload: z.unknown().optional(),
});
export type CanonicalMessage = z.infer<typeof CanonicalMessageSchema>;

// ── Outbound message envelope (what the orchestrator publishes) ─────────

export const OutboundMessageSchema = z.object({
  event_id: z.string().uuid(),
  occurred_at: z.string().datetime(),
  tenant_id: z.string().uuid().nullable(),
  thread_id: z.string().uuid().nullable(),
  target_channel: ChannelSchema,
  target_identifier: z.string(),
  source_phone_number_id: z.string().optional(),
  content: CanonicalContentSchema,
  template_name: z.string().optional(),
  template_variables: z.record(z.string()).optional(),
  requires_window_check: z.boolean().default(true),
  priority: z.enum(['low', 'normal', 'high']).default('normal'),
  trace_context: z.record(z.string()).optional(),
});
export type OutboundMessage = z.infer<typeof OutboundMessageSchema>;

File: packages/python-shared/veda_shared/schemas/canonical.py
from __future__ import annotations

from datetime import datetime
from typing import Annotated, Any, Literal

from pydantic import BaseModel, ConfigDict, Field

from veda_shared.schemas.identity import Channel


class TextContent(BaseModel):
    model_config = ConfigDict(extra="ignore")
    type: Literal["text"]
    text: str = Field(min_length=1, max_length=4096)


class VoiceContent(BaseModel):
    model_config = ConfigDict(extra="ignore")
    type: Literal["voice"]
    media_url: str
    duration_seconds: int | None = None
    transcription: str | None = None
    transcription_language: str | None = None


class ImageContent(BaseModel):
    model_config = ConfigDict(extra="ignore")
    type: Literal["image"]
    media_url: str
    caption: str | None = None
    analysis: str | None = None


class DocumentContent(BaseModel):
    model_config = ConfigDict(extra="ignore")
    type: Literal["document"]
    media_url: str
    filename: str | None = None
    mime_type: str | None = None


class LocationContent(BaseModel):
    model_config = ConfigDict(extra="ignore")
    type: Literal["location"]
    lat: float
    lng: float
    name: str | None = None
    address: str | None = None


class InteractiveReplyContent(BaseModel):
    model_config = ConfigDict(extra="ignore")
    type: Literal["button_reply", "list_reply"]
    selected_id: str
    selected_title: str
    context_message_id: str | None = None


class TemplateComponent(BaseModel):
    type: Literal["header", "body", "button"]
    parameters: list[dict[str, str]] = Field(default_factory=list)


class TemplateContent(BaseModel):
    model_config = ConfigDict(extra="ignore")
    type: Literal["template"]
    template_name: str
    language: str
    components: list[TemplateComponent]


class OutboundButton(BaseModel):
    id: str
    title: str = Field(max_length=20)


class OutboundButtonsContent(BaseModel):
    model_config = ConfigDict(extra="ignore")
    type: Literal["buttons"]
    body_text: str = Field(min_length=1)
    header_text: str | None = None
    footer_text: str | None = None
    buttons: list[OutboundButton] = Field(min_length=1, max_length=3)


class OutboundListItem(BaseModel):
    id: str
    title: str = Field(max_length=24)
    description: str | None = Field(default=None, max_length=72)


class OutboundListSection(BaseModel):
    title: str
    items: list[OutboundListItem] = Field(min_length=1, max_length=10)


class OutboundListContent(BaseModel):
    model_config = ConfigDict(extra="ignore")
    type: Literal["list"]
    body_text: str = Field(min_length=1)
    header_text: str | None = None
    footer_text: str | None = None
    button_text: str = Field(default="Choose", max_length=20)
    list_sections: list[OutboundListSection]


CanonicalContent = Annotated[
    TextContent
    | VoiceContent
    | ImageContent
    | DocumentContent
    | LocationContent
    | InteractiveReplyContent
    | TemplateContent
    | OutboundButtonsContent
    | OutboundListContent,
    Field(discriminator="type"),
]


class DeliveryStatus(str):
    QUEUED = "queued"
    SENT = "sent"
    DELIVERED = "delivered"
    READ = "read"
    FAILED = "failed"


class CanonicalMessageDelivery(BaseModel):
    status: Literal["queued", "sent", "delivered", "read", "failed"]
    error: str | None = None
    sent_at: datetime | None = None
    delivered_at: datetime | None = None
    read_at: datetime | None = None


class CanonicalMessage(BaseModel):
    model_config = ConfigDict(extra="ignore")

    message_id: str
    direction: Literal["inbound", "outbound"]
    channel: Channel
    tenant_id: str | None = None
    thread_id: str | None = None
    sender_principal_id: str | None = None
    sender_identifier: str
    recipient_identifier: str
    timestamp: datetime
    content: CanonicalContent
    delivery: CanonicalMessageDelivery | None = None
    raw_payload: Any | None = None


class OutboundMessage(BaseModel):
    model_config = ConfigDict(extra="ignore")

    event_id: str
    occurred_at: datetime
    tenant_id: str | None = None
    thread_id: str | None = None
    target_channel: Channel
    target_identifier: str
    source_phone_number_id: str | None = None
    content: CanonicalContent
    template_name: str | None = None
    template_variables: dict[str, str] | None = None
    requires_window_check: bool = True
    priority: Literal["low", "normal", "high"] = "normal"
    trace_context: dict[str, str] | None = None
